network/p2p/handler.go

```diff
@@ -0,0 +1,94 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package p2p
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/message"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/utils/logging"
+)
+
+// Handler is the server-side logic for virtual machine application protocols.
+type Handler interface {
+	// AppGossip is called when handling an AppGossip message.
+	AppGossip(
+		ctx context.Context,
+		nodeID ids.NodeID,
+		gossipBytes []byte,
+	) error
+	// AppRequest is called when handling an AppRequest message.
+	// Returns the bytes for the response corresponding to [requestBytes]
+	AppRequest(
+		ctx context.Context,
+		nodeID ids.NodeID,
+		deadline time.Time,
+		requestBytes []byte,
+	) ([]byte, error)
+	// CrossChainAppRequest is called when handling a CrossChainAppRequest
+	// message.
+	// Returns the bytes for the response corresponding to [requestBytes]
+	CrossChainAppRequest(
+		ctx context.Context,
+		chainID ids.ID,
+		deadline time.Time,
+		requestBytes []byte,
+	) ([]byte, error)
+}
+
+// responder automatically sends the response for a given request
+type responder struct {
+	handler Handler
+	log     logging.Logger
+	sender  common.AppSender
+}
+
+func (r responder) AppRequest(ctx context.Context, nodeID ids.NodeID, requestID uint32, deadline time.Time, request []byte) error {
```

Comment:
If this is the protocol specific handler, it would be great to include which protocol this is in the debug logs when encountering an error.